core/src/main/scala/kafka/log/Log.scala

```diff
@@ -304,6 +304,65 @@ class Log(@volatile var dir: File,
       s"log end offset $logEndOffset in ${time.milliseconds() - startMs} ms")
   }
 
+  // the high watermark offset value, in non-leader logs only its message offsets are kept
+  @volatile private[this] var _highWatermarkMetadata : LogOffsetMetadata = LogOffsetMetadata(0)
+
+  def highWatermark: Long = highWatermarkMetadata.messageOffset
+
+  def highWatermark_=(newHighWatermark: Long): Unit = {
+    highWatermarkMetadata = LogOffsetMetadata(newHighWatermark)
+  }
+
+  def highWatermarkMetadata: LogOffsetMetadata = _highWatermarkMetadata
+
+  def highWatermarkMetadata_=(newHighWatermark: LogOffsetMetadata) {
+    if (newHighWatermark.messageOffset < 0)
+      throw new IllegalArgumentException("High watermark offset should be non-negative")
+
+    _highWatermarkMetadata = newHighWatermark
+    onHighWatermarkIncremented(newHighWatermark.messageOffset)
```

Comment:
I think we can either make `onHighWatermarkIncremented` a private method or we can get rid of it. As far as I can tell, the only uses are in test cases and we should probably replace them with one of the methods above to set the high watermark.

Reply:
Done.